### PR TITLE
Aim logger

### PR DESCRIPTION
**Pull request by matdmiller**
Add aim composer logging adapter, aim remote uploader, docker quickstart. Completed several test runs. Requires adding env vars for aim remote upload either via your saas solutions secrets or manually in your own envs. See quickstart for instructions.  The composer logging adapter should get additional reviews and testing but this should be a good start and has gone through several successful test runs.

### CODE REVIEW

**Comment by vishalbakshi** [comment]
@matdmiller I'm working on testing this out. What type of testing would qualify an approving review? A couple training runs of 1000 batches? Or something more involved like the (very rough draft) tests in [my aim-logger-tests branch](https://github.com/LocalResearchGroup/llm-foundry/blob/aim-logger-tests/tests/loggers/test_aim_logger.py)? Want to balance robustness and velocity of testing.

**Comment by galopyz** [comment]
I will work on this one, too. I will try the quickstart and see whether it logs on Mat's logging server.

**Comment by matdmiller** [comment]
I would just do a run with 1k batches or something like that w/ maybe eval every 200 batches or something like that. I do want to do more thorough testing at some point on the adapter specifically and a more detailed code review but I think what I have is a good start and worth merging. Will make it easier for others to incorporate and test once it's merged.

**Comment by galopyz** [comment]
Training works, but I am getting this error:
```
repo_path .aim zip_filepath /llm-foundry/scripts/aim_repo.zip upload_filename aim_repo_20250210_003403.zip zip_filepath.exists() True
2025-02-10 00:34:04,194: rank0[7266][MainThread]: WARNING: llmfoundry.loggers.composer_aim_logger:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!! AIM REMOTE UPLOAD FAILED. PLEASE RE-RUN Manually. !!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Expecting value: line 1 column 1 (char 0)
```

I am running docker locally. 
I set client-Id as my email and secret as the access login thing from the email. 

Does anyone know why I am getting this error?

**Comment by matdmiller** [comment]
Did you set the required environment variables? Are you running locally w/ docker or on Modal? 
You need to run this if you're running in docker:
CleanShot 2025-02-09 at 19 05 47@2x

**Comment by matdmiller** [comment]
Modal Instructions:
Create a `Custom` secret:
CleanShot 2025-02-09 at 19 32 20@2x
Add the 2 secrets below:
CleanShot 2025-02-09 at 19 34 49@2x
Add the secrets to the decorator.
CleanShot 2025-02-09 at 19 36 04@2x

**Comment by vishalbakshi** [comment]
@matdmiller @galopyz I trained SmolLM2-135M with the AIM Logger for 10ba, 100ba, 1000ba (twice), 3000ba and 5000ba (twice) and in all cases the loss goes down. 

For my 5000ba runs:

- I trained with two different learning rates (the default 6e-4 and 6e-3) and the loss curve and final loss values are different (both go down). 
- I set `eval_interval` to 100ba and it logs it and shows different loss curves for the different learning rates (both eventually go down but 6e-3 goes up for a bit before going back down).
- I also cross-referenced the logged "Run Params" with the `smollm2-135m.yaml` file and everything checks out _except_ the following:

In the YAML file it shows:

```
fsdp_config:
  sharding_strategy: FULL_SHARD
  mixed_precision: PURE
  activation_checkpointing: false
  activation_checkpointing_reentrant: false
  activation_cpu_offload: false
  limit_all_gathers: true
```

but in the logged params it shows:

image

and in the YAML it shows:

```
callbacks:
  speed_monitor:
    window_size: 10
  lr_monitor: {}
  memory_monitor: {}
  runtime_estimator: {}
```

but I do not see any `callbacks` params logged by AIM. I don't know enough about FSDP and callbacks to comment on whether this result is okay. Related warning message during training:

```
/llm-foundry/llmfoundry/command_utils/train.py:351: UserWarning: FSDP is not applicable for single-GPU training. Reverting to DDP.
```

My experiment names are `vishal_aim_logger_test_{batch size}`.

Overall these feel like quite good results! Let me know what else we should test.

**Comment by vishalbakshi** [comment]
Let me know if this calculation is correct:

For 5000ba runs I am calculating the total number of tokens trained as:

max seq len x batch size x number of batches = 2048 x 8 x 5000 = ~82M tokens. Without doing eval this took ~30 minutes on 1 A100. With doing eval every 100ba this took ~2 hours. 

IIRC, the very rough rule of thumb was 150M parameter model takes 1 hour to do 1B tokens on a 4090 so this result in comparison seems quite slow.

<mark>**Update**</mark>: I commented out the logger stuff from the YAML file and the training (without doing eval) also took 30 minutes and training with eval every 100ba took ~2 hours so I don't think the logger is at fault.

**Comment by matdmiller** [comment]
Nice! Thanks for testing it out!

> fsdp_config:
>   sharding_strategy: FULL_SHARD

FSDP is for multi GPU only so I'm not too surprised it didn't log anything even though it was in the yaml. I'm not too surprised about the callbacks either.  It's good that everything else was captured!

For the speed I can't remember what he said, but I looked at the modern bert paper and for the pretraining they did 1.7T tokens in 191 hours on a 8xH100 so 1,719,000,000,000/191/8=~1.1B/H100/hr. I looked at the H100 runs I did and they were ~340M tok/hr so not great but also not quite as far off as you were thinking. Between modern bert architecture being faster plus the several 10-20% speedups that he mentioned he implemented that I don't think we have, I'm not too surprised at this result. 

We ready to merge?

**Review by vishalbakshi — APPROVED** [review]
@matdmiller or @galopyz can you merge this PR? I want to make sure members other than me have correct access to merge.

**Comment by matdmiller** [comment]
Yes I can merge but need at least 1 approval first.
